wallet/src/main/java/com/android/identity_credential/wallet/TransferHelper.kt

```diff
@@ -0,0 +1,237 @@
+import com.android.identity.securearea.SecureAreaRepository
+import com.android.identity.storage.StorageEngine
+import com.android.identity.util.Constants
+import com.android.identity.util.Logger
+import java.io.File
+import java.security.KeyPair
+import java.security.PublicKey
+import java.util.OptionalLong
+
+class TransferHelper private constructor(private val context: Context) {
+
+    companion object {
+        private const val TAG = "TransferHelper"
+
+        const val PREFERENCE_KEY_NFC_STATIC_HANDOVER_ENABLED = "nfc_static_handover_enabled"
+        const val PREFERENCE_KEY_NFC_NEGOTIATED_HANDOVER_ENABLED = "nfc_negotiated_handover_enabled"
+
+        const val PREFERENCE_KEY_BLE_CENTRAL_CLIENT_DATA_TRANSFER_ENABLED = "ble_central_client_data_transfer_enabled"
+        const val PREFERENCE_KEY_BLE_PERIPHERAL_SERVER_DATA_TRANSFER_ENABLED = "ble_peripheral_server_data_transfer_enabled"
+        const val PREFERENCE_KEY_WIFI_AWARE_DATA_TRANSFER_ENABLED = "wifi_aware_data_transfer_enabled"
+        const val PREFERENCE_KEY_NFC_DATA_TRANSFER_ENABLED = "nfc_data_transfer_enabled"
+        const val PREFERENCE_KEY_TCP_DATA_TRANSFER_ENABLED = "tcp_data_transfer_enabled"
+        const val PREFERENCE_KEY_UDP_DATA_TRANSFER_ENABLED = "udp_data_transfer_enabled"
+
+        const val PREFERENCE_KEY_L2CAP_ENABLED = "l2cap_enabled"
+        const val PREFERENCE_KEY_DEBUG_ENABLED = "debug_enabled"
+
+        @SuppressLint("StaticFieldLeak")
+        @Volatile
+        private var instance: TransferHelper? = null
+
+        fun getInstance(context: Context) =
+            instance ?: synchronized(this) {
+                instance ?: TransferHelper(context).also { instance = it }
+            }
+    }
+
+    var secureAreaRepository: SecureAreaRepository
+    var androidKeystoreSecureArea: AndroidKeystoreSecureArea
+    var credentialStore: CredentialStore
+
+    private var sharedPreferences: SharedPreferences
+    private var storageEngine: StorageEngine
+    private var deviceRetrievalHelper: DeviceRetrievalHelper? = null
+    private var connectionMethod: ConnectionMethod? = null
+    private var deviceRequest: ByteArray? = null
+
+    private var state = MutableLiveData<State>()
+
+    enum class State {
+        NOT_CONNECTED,
+        ENGAGING,
+        ENGAGEMENT_SENT,
+        CONNECTED,
+        REQUEST_AVAILABLE,
+        RESPONSE_SENT
+    }
+
+    init {
+        val storageDir = File(context.noBackupFilesDir, "identity")
+        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
+        storageEngine = AndroidStorageEngine.Builder(context, storageDir).build()
+        secureAreaRepository = SecureAreaRepository();
+        androidKeystoreSecureArea = AndroidKeystoreSecureArea(context, storageEngine);
+        secureAreaRepository.addImplementation(androidKeystoreSecureArea);
+        credentialStore = CredentialStore(storageEngine, secureAreaRepository)
```

Comment:
These objects are already created in `WalletApplication`, use them instead of duplicating here.